### PR TITLE
Map local build path during WSL run

### LINKED ISSUE #654: Using VS remote test environment results in System.ArgumentNullException : Value cannot be null. (Parameter 'directory')
## Description

Verify throws `System.ArgumentNullException : Value cannot be null. (Parameter 'directory')` when trying to run tests in Visual Studio on Windows using remote test environment targeting  `Ubuntu-20.04` via WSL2.

## How to reproduce

1) Create a new `xUnit` test project.
2) Setup a [remote test environment using WLS2](https://learn.microsoft.com/en-us/visualstudio/test/remote-testing?view=vs-2022#local-wsl2-connections).
3) Create a test that uses `Verify`.
4) Run the test.
5) Test fails with  `System.ArgumentNullException : Value cannot be null. (Parameter 'directory')` exception.

Unless this is not something you are willing to support, I would expect that instead of the exception I would get the usual `Verify` flow generating `received` files and comparing them with `verified` files.

## System

This happens when using the following NuGet package versions:

```
<PackageReference Include="Microsoft.NET.Test.Sdk" Version="17.3.1" />
<PackageReference Include="Verify.Xunit" Version="17.10.2" />
<PackageReference Include="xunit" Version="2.4.2" />
<PackageReference Include="xunit.runner.visualstudio" Version="2.4.5">
<PackageReference Include="coverlet.collector" Version="3.1.2">
<PackageReference Include="Verify.Xunit" Version="17.10.2" />
```

in the following system:

```
.NET SDK (reflecting any global.json):
 Version:   6.0.400
 Commit:    7771abd614

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.22000
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\6.0.400\
```

## More info

I also tested using the latest `Verify` pre-release version: `18.0.0-beta.18`.
It should be straightforward to reproduce this issue using the steps outlined above but if it helps, I could create a repo to reproduce this.

### testenvironments.json

This was the test environment file I used:

```json
{
  "version": "1",
  "environments": [
    {
      "name": "WSL-Ubuntu-20.04",
      "type": "wsl",
      "wslDistribution": "Ubuntu-20.04"
    }
  ]
}
```

### Verify 17.10.2 exception stack trace

```
  Message: 
    System.ArgumentNullException : Value cannot be null. (Parameter 'directory')

  Stack Trace: 
    Guard.AgainstEmpty(String value, String argumentName) line 99
    Guard.BadDirectoryName(String name, String argumentName) line 50
    PathInfo.ctor(String directory, String typeName, String methodName) line 15
    <.cctor>b__42_2(String sourceFile, String projectDirectory, Type type, MethodInfo method) line 23
    VerifierSettings.GetPathInfo(String sourceFile, Type type, MethodInfo method) line 29
    ReflectionFileNameBuilder.FileNamePrefix(MethodInfo method, Type type, String sourceFile, VerifySettings settings, String uniquenessForReceived, String uniquenessForVerified) line 11
    <>c__DisplayClass0_0.<GetVerifier>b__0(String uniquenessForReceived, String uniquenessForVerified) line 16
    InnerVerifier.ctor(String sourceFile, VerifySettings settings, GetFileConvention fileConvention) line 17
    Verifier.GetVerifier(VerifySettings settings, String sourceFile) line 12
    <<Verify>b__0>d.MoveNext() line 30
    --- End of stack trace from previous location ---
    ReadDataFromDifferentGitHubWorkflowCommandDependencyErrorTests.FailedHttpToDownloadArtifact() line 102
    --- End of stack trace from previous location ---
```

### Verify 18.0.0-beta.18 exception stack trace

```
  Message: 
    System.ArgumentNullException : Value cannot be null. (Parameter 'directory')

  Stack Trace: 
    Guard.AgainstEmpty(String value, String argumentName) line 108
    Guard.BadDirectoryName(String name, String argumentName) line 59
    PathInfo.ctor(String directory, String typeName, String methodName) line 15
    <.cctor>b__26_0(String sourceFile, String projectDirectory, Type type, MethodInfo method) line 10
    Verifier.GetPathInfo(String sourceFile, Type type, MethodInfo method) line 18
    Verifier.GetVerifier(VerifySettings settings, String sourceFile, Boolean useUniqueDirectory) line 23
    <<Verify>b__0>d.MoveNext() line 63
    --- End of stack trace from previous location ---
    ReadDataFromDifferentGitHubWorkflowCommandDependencyErrorTests.FailedHttpToDownloadArtifact() line 102
    --- End of stack trace from previous location ---
```

### PR DESCRIPTION
**Pull request by JanKrivanek**
## Problem

Fixes #654
Running under WSL crashes due to mixing windows and subsystem mapped files
Related discussion within https://github.com/VerifyTests/Verify/pull/700

## Proposed solution

Added prototype mapper function mapping the build time paths to mapped WSL paths.
Used the mapping 'as late as possible' - so with exception of `Expecto` (that needs source path to infer calling type) it's in the `ResolveDirectory`, plus in `AttributeReader` (as that's publicly exposed)

## What is missing

Testing - unit testing is bit problematic with this type of feature

### CODE REVIEW

**Comment by SimonCropp** [comment]
is it possible to add a test for this scenario? either locally or on CI?

**Comment by JanKrivanek** [comment]
> is it possible to add a test for this scenario? either locally or on CI?

Would it be acceptable to have manual testing for this?
I added `testenvironments.json` to be able to localy run unit tests on WSL:

image

(note some of the tests failing under WSL - this is most likely mostly due to mismatched matching of streams between environemnts. This is not related to the way how the WSL is supported).


There can likely be an additional CI job, which setups additional docker container and runs tests there - I unfortunately cannot commit to this work during this year.

**Comment by SimonCropp** [comment]
> Would it be acceptable to have manual testing for this?
> I added testenvironments.json to be able to localy run unit tests on WSL:

yep thats sufficient. can you add a section the readme on how to do that test. then i will merge and ship this

**Comment by JanKrivanek** [comment]
@SimonCropp I've reworked a bit - so one more look might be needed.
I wanted to make dockerized runs supported as well - so that it can be prepared for CI testing scenario in the future (I'm trying to see if this feature can be used from CLI and will comment back here or create item once I have some details).

I as well created separate item for fixing the tests failing for those scenarios.

**Comment by JanKrivanek** [comment]
https://github.com/dotnet/sdk/issues/29516

**Comment by JanKrivanek** [comment]
@SimonCropp - This feels ready from mine point of view. Though I'm very open to any suggestions/ideas.

File: src/Verify/IoHelpers.cs [hunk 1]
```diff
@@ -203,6 +203,24 @@ public static string GetRelativePath(string directory, string file)
...
     }
 
+    static VirtualizedRunHelper? virtualizedRunHelper;
+
+    internal static void MapPathsForCallingAssembly(Assembly assembly) =>
+        virtualizedRunHelper = new(assembly);
+
+    internal static string? GetMappedBuildPath(string? path, Assembly? assembly = null)
+    {
+        if (virtualizedRunHelper == null && assembly != null)
```
Comment (SimonCropp) [inline src/Verify/IoHelpers.cs:213]: this is problematic since based on people use of `AttributeReader` the `virtualizedRunHelper ` can be mutated. shouldnt the static instance of `virtualizedRunHelper` only be initialized in `TargetAssembly.Assign`?

Comment (JanKrivanek) [reply src/Verify/IoHelpers.cs]: Correct - I originally intentionaly had it that way - as I was expecting all the callers to `AttributeReader` and `Verify` to be located in a code that is all going through single build in a single location. While the code can be in multiple assemblies, it still would have single root (solution) - so despite performing mapping via different projects/assemblies, result would be same.
Would the code be comming from different builds - then hopes for proper mapping of every single path are lost due to inability to locate and reach the other build code location from the virtualized mapped path.

...

At least that was my thinking.. but based on your comment I realized there is one specific scenario where separating mappers per assembly makes sense - and that's when the callers are in different assemblies, comming from different builds, but all those builds occured on the same developer machine and from same drive. The code roots might differ. All the code is still reachable from the VM/docker then and the mapping code should be able to reflect that.

A bit esoteric scenario (yeah - I'm the last one to speak about esoteric scenarios), but not nonexistent, plus code is cleaner this way. So thanks for detailed look!

**tl;dr;:** **I have updated the code to separate mappers per assembly.**

File: src/Verify/VirtualizedRunHelper.cs [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+            return;
+        }
+
+        string currentDir = Environment.CurrentDirectory;
+        string currentDirRelativeToAppRoot = currentDir.TrimStart(Separators);
+
+        // WSL paths mount to /mnt/<drive>/...
+        // docker testing mounts to /mnt/approot/...
+        if (!TryRemoveDirFromStartOfPath(ref currentDirRelativeToAppRoot) ||
+            !TryRemoveDirFromStartOfPath(ref currentDirRelativeToAppRoot))
```
Comment (SimonCropp) [inline src/Verify/VirtualizedRunHelper.cs:30]: why is this line duplicated?

Comment (SimonCropp) [reply src/Verify/VirtualizedRunHelper.cs:30]: ahh i see. each time it modifies the ref param.

Comment (JanKrivanek) [reply src/Verify/VirtualizedRunHelper.cs:30]: This could have surely been better commented :-|
